test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/JSOptionalTest.scala

```diff
@@ -0,0 +1,291 @@
+    trait TraitWithAbstracts extends js.Object {
+      val x: js.UndefOr[Int]
+      def y: js.UndefOr[String]
+      def y2: js.UndefOr[String]
+      var z: js.UndefOr[Option[Int]]
+    }
+
+    @ScalaJSDefined
+    trait OverrideTraitAbstractWithOptional extends TraitWithAbstracts {
+      val x: js.UndefOr[Int] = js.undefined
+      def y: js.UndefOr[String] = js.undefined
+      val y2: js.UndefOr[String] = js.undefined
+      var z: js.UndefOr[Option[Int]] = js.undefined
+    }
+
+    val obj = new OverrideTraitAbstractWithOptional {}
+
+    assertEquals(js.undefined, obj.x)
+    assertFalse(obj.hasOwnProperty("x"))
+
+    assertEquals(js.undefined, obj.y)
+    assertFalse(js.Object.hasProperty(obj, "y"))
+
+    assertEquals(js.undefined, obj.y2)
+    assertFalse(js.Object.hasProperty(obj, "y2"))
+
+    assertEquals(js.undefined, obj.z)
+    assertFalse(obj.hasOwnProperty("z"))
+    obj.z = Some(3)
+    assertEquals(Some(3), obj.z)
+  }
+
+  def polyOptionalMethod(): Unit = {
+    @ScalaJSDefined
+    trait TraitWithPolyOptionalMethod extends js.Object {
+      def foo[A]: js.UndefOr[A] = js.undefined
+    }
+
+    val obj = new TraitWithPolyOptionalMethod {}
+
+    assertEquals(js.undefined, obj.foo[Int])
+    assertFalse(js.Object.hasProperty(obj, "foo"))
+  }
+}
+
+object JSOptionalTest {
+  @ScalaJSDefined
+  trait TraitWithOptional extends js.Object {
+    val x: js.UndefOr[Int] = js.undefined
+    def y: js.UndefOr[String] = js.undefined
+    def y2: js.UndefOr[String] = js.undefined
+    var z: js.UndefOr[Option[Int]] = js.undefined
+  }
+
+  @ScalaJSDefined
+  class ClassImplementsTraitWithOptional extends TraitWithOptional
+
+  @ScalaJSDefined
+  class UndefinedInClassIsNotOptional extends js.Object {
+    val x: js.UndefOr[Int] = js.undefined
+    def y: js.UndefOr[String] = js.undefined
+    def y2: js.UndefOr[String] = js.undefined
+    var z: js.UndefOr[Option[Int]] = js.undefined
+  }
+
+  @ScalaJSDefined
+  class OverrideWithUndefinedInClassIsNotOptional extends TraitWithOptional {
+    override val x: js.UndefOr[Int] = js.undefined
+    override def y: js.UndefOr[String] = js.undefined
+    override def y2: js.UndefOr[String] = js.undefined
+    z = js.undefined
+  }
+
+  @ScalaJSDefined
+  class ClassImplementsTraitWithOptionalOverrideWithConcrete
+      extends TraitWithOptional {
+    override val x: js.UndefOr[Int] = 42
+    override val y: js.UndefOr[String] = "hello"
+    override def y2: js.UndefOr[String] = "world"
+    z = Some(5)
+  }
+
+  @ScalaJSDefined
+  trait OverrideOptionalWithOptional extends TraitWithOptional {
+    override val x: js.UndefOr[Int] = js.undefined
+    override val y: js.UndefOr[String] = js.undefined
+    override def y2: js.UndefOr[String] = js.undefined
+  }
+
+  @ScalaJSDefined
+  trait OverrideOptionalWithOptionalImplicitType extends TraitWithOptional {
+    override val x = js.undefined
+    override val y = js.undefined
+    override def y2 = js.undefined // scalastyle:ignore
```

Comment:
Because `js.undefined` has type `js.UndefOr[Nothing]`, which is actually a subtype of `js.UndefOr[String]`, so it works because `y2` is redefined with a more specific type. Other examples do not work because `String` is not actually a subtype of `js.UndefOr[String]` (it needs an implicit conversion).